### ISSUE TITLE
Adding a .projectile file causes "projectile-invalidate-cache: Wrong type argument: hash-table-p, nil"

### ISSUE THREAD

**Issue by jwr** [bbatsov/projectile#1854]
## Expected behavior

Adding a .projectile file to the root of a project should not stop `projectile-find-file` from working.

## Actual behavior

Adding a .projectile file (either empty or containing excluded directories) to the root of a project causes `projectile-find-file` to fail with "Wrong type argument: hash-table-p, nil"

## Steps to reproduce the problem

Add a .projectile file
run `projectile-find-file`

### Projectile version information

`Projectile 2.7.0`

### Emacs version

Emacs 29.1.50, earlier also on Emacs 28

### Operating system

Mac OS 12.6.7

**Comment by Pacane** [comment]
I'm also hitting this issue. Do you have any workaround so far @jwr ?

**Comment by jwr** [comment]
Unfortunately, no. I learned to live with it and the pain of `projectile-find-file` listing a ton of irrelevant stuff.

**Comment by jwr** [comment]
This issue still exists for me. Emacs 29.1, Projectile 2.8.0, macOS 15.1.1.

I sadly cannot use a `.projectile` file to limit what gets considered for opening.

**Comment by jwr** [comment]
Here is a backtrace:

```
Debugger entered--Lisp error: (wrong-type-argument hash-table-p nil)
  remhash("/Users/jwr/Sync/code/" nil)
  projectile-invalidate-cache(nil)
  projectile-maybe-invalidate-cache(nil)
  (if (and (eq projectile-require-project-root 'prompt) (not (projectile-project-p))) (counsel-projectile-find-file-action-switch-project) (projectile-maybe-invalidate-cache arg) (+projectile-find-file--initialize) (let* ((project-files (+projectile-find-file--sort-file-with-scores (projectile-current-project-files))) (filename (completing-read (projectile-prepend-project-name "Find file: ") project-files nil t))) (if filename (progn (find-file (projectile-expand-root filename)))) (if nil (progn (ivy-read (projectile-prepend-project-name "Find file: ") project-files :matcher #'+projectile-find-file--matcher :require-match t :action counsel-projectile-find-file-action :caller '+projectile-find-file)))))
  +projectile-find-file(nil)
  funcall-interactively(+projectile-find-file nil)
  call-interactively(+projectile-find-file nil nil)
  command-execute(+projectile-find-file)
```

**Comment by bbatsov** [comment]
The problem seems to be triggered by:

```
(defun projectile-maybe-invalidate-cache (force)
  "Invalidate if FORCE or project's dirconfig newer than cache."
  (when (or force (file-newer-than-file-p (projectile-dirconfig-file)
                                          projectile-cache-file))
    (projectile-invalidate-cache nil)))
```

I'm puzzled how this happens in your case, though, as without a cache file this should never enter the `projectile-invalidate-cache` command when the error occurs.

**Comment by bbatsov** [comment]
I think I have a pretty good guess as to what's going on here. Would I be right if I assumed you're not using `projectile-mode`? I'm asking because I see some cache init logic is there and I'm guessing things break for people not using `projectile-mode` directly. Should be an easy fix, provided that's the root cause.

**Comment by jwr** [comment]
First, my apologies for an egg-on-face moment: what I initially reported was *not* about `projectile-find-file`, but about `+projectile-find-file`, which is part of a horrible hack to the default ranking of files in projectile. I've been using that code for years, as it vastly improves my experience with projectile-find-file. Notice the `+` in the backtraces above…

However.

When I use the actual `projectile-find-file` I can also reproduce the error. With your hint, I was able to debug and find out why `projectile-invalidate-cache` is called. It's because if I create a `.projectile` file (empty) in the current project, `file-newer-than-file-p` gets called with that `.projectile` file and `projectile-cache-file` which is `~/.emacs.d/projectile.cache` which doesn't exist in my filesystem. `file-newer-than-file-p` returns `t` in this case, which causes the call to `projectile-invalidate-cache`.

If I `touch ~/.emacs.d/projectile.cache`, the problem disappears.

In other words: my initial report was garbage, but the problem still exists, even though I am not using the cache.

**Comment by jwr** [comment]
> I think I have a pretty good guess as to what's going on here. Would I be right if I assumed you're not using projectile-mode?

That… is a trickier question than I imagined.

I never explicitly call `projectile-mode` and `projectile-mode` variable is nil.

but this is how I initialize projectile:

```
(use-package projectile
  :config
  (setq projectile-global-mode t
		projectile-current-dir-priority t
		projectile-sort-order 'modification-time))
```

**Comment by bbatsov** [comment]
I did some changes to the caching logic in https://github.com/bbatsov/projectile/commit/5061bd8dcd9f4d0e874884272f88b10892d15da3 that I believe should solve your problem (amongst other problems).

**Comment by bbatsov** [comment]
At any rate - I think my last round of changes should have solved the problem for you.

**Comment by jwr** [comment]
I can confirm that projectile 2.9.1 fixes the problem! Thank you! 🙏
